st2api/tests/unit/controllers/exp/test_inquiries.py

```diff
@@ -0,0 +1,385 @@
+        self.assertEqual(put_resp.status_int, http_client.BAD_REQUEST)
+        self.assertIn('is not an Inquiry', put_resp.json['faultstring'])
+
+    @mock.patch('st2api.controllers.exp.inquiries.action_service')
+    def test_respond_no_parent(self, mock_as):
+        """Test that a resume was not requested for an Inquiry without a parent
+        """
+
+        post_resp = self._do_create_inquiry(INQUIRY_2, RESULT_DEFAULT)
+        inquiry_id = self._get_inquiry_id(post_resp)
+        response = {"continue": True}
+        put_resp = self._do_respond(inquiry_id, response)
+        self.assertEqual(response, put_resp.json.get("response"))
+        mock_as.request_resume.assert_not_called()
+
+    def test_respond_duplicate_rejected(self):
+        """Test that responding to an already-responded Inquiry fails
+        """
+
+        post_resp = self._do_create_inquiry(INQUIRY_2, RESULT_DEFAULT)
+        inquiry_id = self._get_inquiry_id(post_resp)
+        response = {"continue": True}
+        put_resp = self._do_respond(inquiry_id, response)
+        self.assertEqual(response, put_resp.json.get("response"))
+
+        # The inquiry no longer exists, since the status should not be "pending"
+        # Get the execution and confirm this.
+        inquiry_execution = self._do_get_execution(inquiry_id)
+        self.assertEqual(inquiry_execution.json.get('status'), 'succeeded')
+
+        # A second, equivalent response attempt should not succeed, since the Inquiry
+        # has already been successfully responded to
+        put_resp = self._do_respond(inquiry_id, response, expect_errors=True)
+        self.assertEqual(put_resp.status_int, http_client.BAD_REQUEST)
+        self.assertIn('has already been responded to', put_resp.json['faultstring'])
+
+    def test_respond_restrict_users(self):
+        """Test that Inquiries can reject responses from users not in a list
+        """
+
+        # Default user for tests is "stanley", which is not in the 'users' list
+        # Should be rejected
+        post_resp = self._do_create_inquiry(INQUIRY_2, RESULT_2)
+        inquiry_id = self._get_inquiry_id(post_resp)
+        response = {"continue": True}
+        put_resp = self._do_respond(inquiry_id, response, expect_errors=True)
+        self.assertEqual(put_resp.status_int, http_client.FORBIDDEN)
+        self.assertIn('Insufficient permission to respond based on Inquiry parameters.',
+                      put_resp.json['faultstring'])
+
+        # Responding as a use in the list should be accepted
+        old_user = cfg.CONF.system_user.user
+        cfg.CONF.system_user.user = "foo"
+        post_resp = self._do_create_inquiry(INQUIRY_2, RESULT_2)
+        inquiry_id = self._get_inquiry_id(post_resp)
+        response = {"continue": True}
+        put_resp = self._do_respond(inquiry_id, response)
+        self.assertEqual(response, put_resp.json.get("response"))
+
+        # Clean up
+        cfg.CONF.system_user.user = old_user
```

Comment:
test respond to expired inquiry?

Reply:
Among other things, I added a test for this in https://github.com/StackStorm/st2/commit/577d019f0ebd54af33d22a8d108674710e9e8748